### PR TITLE
Escape Backslashes in properties when using secrets protection && Mask More Sensitive Properties

### PR DESCRIPTION
**Pull request by domenicbove**
# Description

- Secrets protection CLI has the [limitation that all backslashes need to be escaped by another backslash](https://docs.confluent.io/platform/current/security/secrets.html#limitations)
- Additionally users expect more than just properties containing 'password' to get masked. cp-ansible can intelligently mask more!

Technical changes:

- When creating the initial unmasked properties file, write it with escaped backslashes. This is done by temporarily updating the properties dictionary to have 2x the number of backslashes using the `regex_replace` filter.
- Created a `properties.j2` template within the common role. This was useful in creating the unmasked props file, but later down the line, we can remove all duplicates of the properties.j2 template in each component role and just pass a variable to this template as done here.
- Beefed up our default encrypt properties logic. Now all properties matching this regex pattern list will get masked by default:
```
      - '.*password.*'
      - '.*basic.auth.user.info.*'
      - '^ldap.java.naming.security.credentials$'
      - '^confluent.license$'
      - '.*sasl.jaas.config'
```
- Combines the default encrypt properties list with customer provided encrypt properties list (with a unique filter applied of course) to create the `final_encrypt_properties` list of properties to encrypt. 
- With the `final_encrypt_properties` generated, can now call `confluent secret` command **only once** instead of multiple times with a loop. This improves efficiency as well as helps with the backslash limitation. Each command run will cut away a backslash. So this is important to only run once.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Updated `mtls-custombundle-rhel` and `rbac-mtls-rhel` tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 1]
```diff
@@ -51,25 +61,32 @@
...
 
-- name: Encrypt All Properties that Contain 'password'
-  shell: |
-    {{ confluent_cli_path }} secret file encrypt --config-file {{ config_path }} \
-      --local-secrets-file {{ secrets_file }} \
-      --remote-secrets-file {{ secrets_file }}
-  environment:
-    CONFLUENT_SECURITY_MASTER_KEY: "{{ secrets_protection_masterkey }}"
-  when: encrypt_passwords|bool
+- name: Create Encyrpt Properties List
```
Comment (erikgb) [inline roles/confluent.common/tasks/secrets_protection.yml:64]:
```suggestion
- name: Create Encrypt Properties List
```

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 2]
```diff
@@ -51,25 +61,32 @@
...
+  set_fact:
+    encryrpt_properties: "{{ final_properties | dict2items | map(attribute='key') | select('match', properties_patterns|join('|'))
```
Comment (erikgb) [inline roles/confluent.common/tasks/secrets_protection.yml:66]:
```suggestion
    encrypt_properties: "{{ final_properties | dict2items | map(attribute='key') | select('match', properties_patterns|join('|'))
```

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 3]
```diff
@@ -51,25 +61,32 @@
...
+    properties_patterns:
+      - '.*password.*'
+      - '.*basic.auth.user.info.*'
+      - 'ldap.java.naming.security.credentials'
+      - 'confluent.license'
+      - '.*sasl.jaas.config'
+
+- name: Debug Properties to Mask
+  debug:
+    var: encryrpt_properties
```
Comment (erikgb) [inline roles/confluent.common/tasks/secrets_protection.yml:78]:
```suggestion
    var: encrypt_properties
```

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 4]
```diff
@@ -51,25 +61,32 @@
...
-      --remote-secrets-file {{ secrets_file }}
-  environment:
-    CONFLUENT_SECURITY_MASTER_KEY: "{{ secrets_protection_masterkey }}"
-  when: encrypt_passwords|bool
+- name: Create Encyrpt Properties List
+  set_fact:
+    encryrpt_properties: "{{ final_properties | dict2items | map(attribute='key') | select('match', properties_patterns|join('|'))
+      + properties if encrypt_passwords|bool else properties }}"
+  vars:
+    properties_patterns:
```
Comment (erikgb) [inline roles/confluent.common/tasks/secrets_protection.yml:69]: Reading the PR description, I get the impression that the user should be allowed to modify this list. Aren't you missing a variable for that in `confluent.variables` role?

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 5]
```diff
@@ -51,25 +61,32 @@
...
 
-- name: Encrypt Selected Properties
+- name: Encrypt Properties
   shell: |
     {{ confluent_cli_path }} secret file encrypt --config-file {{ config_path }} \
       --local-secrets-file {{ secrets_file }} \
       --remote-secrets-file {{ secrets_file }} \
-      --config {{ item }}
-  loop: "{{ properties }}"
+      --config "{{ encryrpt_properties | join (',') }}"
```
Comment (erikgb) [inline roles/confluent.common/tasks/secrets_protection.yml:86]:
```suggestion
      --config "{{ encrypt_properties | join (',') }}"
```

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 6]
```diff
@@ -51,25 +61,32 @@
...
   environment:
     CONFLUENT_SECURITY_MASTER_KEY: "{{ secrets_protection_masterkey }}"
-  when: properties|length>0
+  changed_when: true
```
Comment (erikgb) [inline roles/confluent.common/tasks/secrets_protection.yml:89]: This looks a bit bogus - at least without a `when`. Are you able to detect if the properties are already encrypted and then skip the task? At least I prefer idempotent playbooks - where everything is green (no changes) on second run. Or am I missing something?

**Review by erikgb — CHANGES_REQUESTED** [review]
I did an initial scan, but I surely have to look closer at the details....

File: roles/confluent.common/tasks/secrets_protection.yml [hunk 7]
```diff
@@ -22,6 +28,10 @@
     owner: "{{secrets_file_owner}}"
     group: "{{secrets_file_group}}"
```
Comment (domenicbove) [inline roles/confluent.common/tasks/secrets_protection.yml:29]: @erikgb This is where I came up with the whole backup/restore idea lol! I should test if this works with a `var`

File: roles/confluent.common/tasks/secrets_protection.yml (hunk 4 above)
Comment (domenicbove) [reply roles/confluent.common/tasks/secrets_protection.yml:69]: ya so you get two variables
```
kafka_broker_secrets_protection_encrypt_passwords: true/false # defaults to true
kafka_broker_secrets_protection_encrypt_properties: [list,of,properties] # defaults to []
```

So if you say encrypt_passwords, then the list gets everything in the pattern list + the user provided property list (which may be empty)

If you don't say encrypt_passwords then this list will only have your user provided property list.

Thinking about it now, I should apply the unique filter after combining the two lists

Comment (erikgb) [reply roles/confluent.common/tasks/secrets_protection.yml:69]: But what if you don't want to encrypt some properties matching the "default patterns"? Wouldn't it be better to have a default, and allow for a complete override?

File: roles/confluent.common/tasks/secrets_protection.yml (hunk 6 above)
Comment (domenicbove) [reply roles/confluent.common/tasks/secrets_protection.yml:89]: So unfortunately this secrets protection logic is not very idempotent bc I need to reset the server.properties file with unmasked props before running this script which then rewrites the file.

But i could add back the when condition bc i suppose someone could say
```
kafka_broker_secrets_protection_encrypt_passwords: false
kafka_broker_secrets_protection_encrypt_properties: []
```
and in that case the `properties|length==0`

Also if you can notice in the next copy task I had to do some fancy logic around triggering the handler. Essentially i compare the old masked file with the new one and if theyre different then trigger the restart.

File: roles/confluent.common/tasks/secrets_protection.yml (hunk 4 above)
Comment (domenicbove) [reply roles/confluent.common/tasks/secrets_protection.yml:69]: if you dont want to encrypt the properties matching the "default patters" just set 
```
kafka_broker_secrets_protection_encrypt_passwords: false
```
and set
```
kafka_broker_secrets_protection_encrypt_properties: [the, props, you, do, want]
```

Comment (erikgb) [reply roles/confluent.common/tasks/secrets_protection.yml:69]: Hmmm, isn't that a bit confusing? 😀 I would expect "encrypt properties" to control this. "Encrypt password" is a bit unprecise to me. The term password could mean a lot of things. IMO the term secret is better. https://martinfowler.com/bliki/TwoHardThings.html

Comment (domenicbove) [reply roles/confluent.common/tasks/secrets_protection.yml:69]: ya, so for more context. its super easy to turn secrets protection on. and with the initial secrets protection release, if you turned it on you'd get all props containing password masked. At the time, I knew that was insufficient, and gave the additional option to add on top of that.

Now i'm trying to make things better by masking more by default (hence the pattern list)

I also don't want to introduce new variables... so i just updated the docs. I do agree the var names aren't the best haha

I feel the best user experience is you just turn it on and dont have to include any additional props bc cp-ansible is smart enough to mask the correct props. This pr is heading in that direction.

**Pull request merged**